src/backend/metal/src/command.rs

```diff
@@ -113,128 +259,386 @@ struct CommandBufferInner {
+                passes.clear();
+                if release_resources {
+                    passes.shrink_to_fit();
+                }
             }
-        }
+        };
+        self.reset_resources();
     }
 
-    fn begin_render_pass(&mut self, encoder: metal::RenderCommandEncoder) {
-        self.stop_ecoding();
+    fn stop_encoding(&mut self) {
+        match self.sink {
+            CommandSink::Immediate { ref mut encoder_state, .. } => {
+                match mem::replace(encoder_state, EncoderState::None)  {
+                    EncoderState::None => {}
+                    EncoderState::Render(ref encoder) => {
+                        encoder.end_encoding();
+                    }
+                    EncoderState::Blit(ref encoder) => {
+                        encoder.end_encoding();
+                    }
+                    EncoderState::Compute(ref encoder) => {
+                        encoder.end_encoding();
+                    }
+                }
+            }
+            CommandSink::Deferred { ref mut is_encoding, .. } => {
+                *is_encoding = false;
+            }
+        }
+    }
 
-        self.encoder_state = EncoderState::Render(encoder);
-        let encoder = if let EncoderState::Render(ref encoder) = self.encoder_state {
-            encoder
-        } else {
-            unreachable!()
-        };
+    fn begin_render_pass(&mut self, descriptor: metal::RenderPassDescriptor) {
+        self.stop_encoding();
 
+        // TODO: re-use storage
+        let mut commands = Vec::new();
         // Apply previously bound values for this command buffer
-        if let Some(viewport) = self.viewport {
-            encoder.set_viewport(viewport);
+        commands.extend(self.viewport.map(soft::RenderCommand::SetViewport));
+        commands.extend(self.scissors.map(soft::RenderCommand::SetScissor));
+        let depth_stencil = self.depth_stencil_state.clone();
+        commands.extend(self.render_pso.clone().map(|pipeline| {
+            soft::RenderCommand::BindPipeline(pipeline, depth_stencil)
+        }));
+        let stages = [pso::Stage::Vertex, pso::Stage::Fragment];
+        for (&stage, resources) in stages.iter().zip(&[&self.resources_vs, &self.resources_fs]) {
+            commands.extend(resources.buffers.iter().enumerate().filter_map(|(i, resource)| {
+                resource.clone().map(|(buffer, offset)| {
+                    soft::RenderCommand::BindBuffer {
+                        stage,
+                        index: i as _,
+                        buffer: Some(buffer),
+                        offset,
+                    }
+                })
+            }));
+            commands.extend(resources.textures
+                .iter()
+                .cloned()
+                .enumerate()
+                .filter(|&(_, ref resource)| resource.is_some())
+                .map(|(i, texture)| soft::RenderCommand::BindTexture {
```

Comment:
Probably could do `filter_map` like the above for these cases (and cloning lazily), although this will change when we reuse storage anyway